### PR TITLE
✨AI to doc editor

### LINKED ISSUE #260: ✨Integrate IA feature
## Feature Request

This branch has the feature integrated: https://github.com/numerique-gouv/impress/tree/feature/ai-blocknote-back

## To finalize

- [x] Clean code (endpoint backend out of document)
- [x] Add e2e test case
- [x] Update helm secret with the ai pass

### PR DESCRIPTION
**Pull request by AntoLC**
## Purpose

Add AI actions to the doc editor: 
- prompt (use the text selected as a prompt)
- correct (fix typo)
- rephrase
- summarize
- translate_en
- translate_de
- translate_fr

We can quite easily add more actions.

## Proposal

- [x] ✨(backend) create ai endpoint
- [x] ✨(frontend) add ai blocknote feature
- [x] 🔧(helm) add ai setting to environments
- [x]  tests

## Demo

[scrnli_9_23_2024_9-52-03 AM.webm](https://github.com/user-attachments/assets/1c30931f-811b-4121-9313-1e28cd503a63)

### CODE REVIEW

File: src/backend/core/api/viewsets.py [hunk 1]
```diff
@@ -303,6 +305,23 @@ def perform_update(self, serializer):
...
+
+    def determine_metadata(self, request, view):
+        """Add language choices only for the list endpoint."""
+        simple_metadata = super().determine_metadata(request, view)
+
+        if request.path.endswith("/documents/"):
+            simple_metadata["actions"]["POST"]["language"] = {
+                "choices": [
+                    {"value": code, "display_name": name}
+                    for code, name in enums.ALL_LANGUAGES.items()
```
[the remark this answers is not in the document]
Comment (AntoLC) [reply src/backend/core/api/viewsets.py:319]: Quite impressive:

[scrnli_fpGawx7SFVKHHU.webm](https://github.com/user-attachments/assets/e0bde35e-9476-4da5-96b4-2fa8bc2619cb)

image

[the remark this answers is not in the document]
Comment (AntoLC) [reply src/backend/core/api/viewsets.py:319]: About the endpoint `OPTIONS documents`
Could we return only the languages ? 
The response that we get is quite big, not sure all these data are usefull:
 
Capture d'écran 2024-10-14 124214

[the remark this answers is not in the document]
Comment (sampaccoud) [reply src/backend/core/api/viewsets.py:319]: It's not returned by me it's part of DRF. I just added the language and the rest was already there...

File: src/backend/core/services/ai_services.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        content = response.choices[0].message.content
+        sanitized_content = re.sub(r"(?<!\\)\n", "\\\\n", content)
+        sanitized_content = re.sub(r"(?<!\\)\t", "\\\\t", sanitized_content)
+
+        json_response = json.loads(sanitized_content)
+
+        if "answer" not in json_response:
+            raise RuntimeError("AI response does not contain an answer")
+
+        return json_response
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply src/backend/core/services/ai_services.py:103]: Are you questioning the quality of ChatGPT 's output? I've always been told that actorization only starts after 3 repeats.
Ok I will factorize the obvious one only for now.

File: src/backend/core/api/utils.py [hunk 3]
```diff
@@ -31,3 +36,135 @@ def generate_s3_authorization_headers(key):
...
+            return 86400  # Throttled by day limit, wait 24 hours
+        if self.recent_requests_hour >= self.rates["hour"]:
+            return 3600  # Throttled by hour limit, wait 1 hour
+        if self.recent_requests_minute >= self.rates["minute"]:
+            return 60  # Throttled by minute limit, wait 1 minute
+
+        return None  # No backoff required
+
+
+class AIUserRateThrottle(BaseThrottle):
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply src/backend/core/api/utils.py:98]: prompt: Make it DRY: {then copy/paste the 2 classes}

Replace in my code
Play tests
git add --all
git commit ^^

**Review by qbey — APPROVED** [review]

File: src/backend/core/services/ai_services.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AntoLC) [reply src/backend/core/services/ai_services.py]: Ah.. mister GPT 😬

**Pull request merged**